File: 2022/c/03_bags/bags.c
#include <stdio.h>
#include <string.h>
#include <ctype.h>
#include <stdlib.h>

int getScore(const char n);
char search(const char arr[]);

int main() {
    FILE* file = NULL;
    char line[50];
    int total = 0;

    if (!(file = fopen("input.txt", "r"))) {
        puts("FILE DID NOT OPEN");
        return 1;
    }

    while (fgets(line, sizeof(line), file)) {
        const char found = search(line);
        const int len = strlen(line);
        printf("\nline: ");
        for (size_t i = 0; i < len; i++) {
            if (line[i] == found) {
                printf("(%c)", line[i]);
            } else {
                printf("%c", line[i]);
            }
        }
        printf("found char: %c\n", found);
        const int s = getScore(found);
        printf("score: %i\n", s);
        total += s;
    }

    printf("\nTotal: %i", total);
}

char search(const char arr[]) {
    size_t len = strlen(arr) - 1;
    size_t mid = len / 2;

    // split
    char *a = malloc((mid + 1) * sizeof(char));
    char *b = malloc((len - mid + 1) * sizeof(char));

    strncpy(a, arr, mid);
    strncpy(b, arr + mid, len - mid + 1);

    // search
    for (size_t i = 0; i < mid + 1; i++) {
        char c = a[i];
        for (size_t j = 0; j < len - mid + 1; j++) {
            if (a[i] == b[j]) {
                free(a);
                free(b);
                return c;
            }
        }
    }

    free(a);
    free(b);
    return 0;
}

int getScore(char found) {
    if (islower(found)) {
        return found - 96;
    }
    return found - 38;
}

File: 2022/go/03_bags/bags.go
package main

import (
    "bufio"
    "fmt"
    "os"
    "unicode"
)

func main() {
    file, err := os.Open("input")
    if err != nil {
        fmt.Fprintf(os.Stderr, "%s", err)
    }
    defer file.Close()

    buf := bufio.NewScanner(file)

    total := 0
    for buf.Scan() {
        line := buf.Text()
        found := search(line)

        fmt.Printf("line: ")
        for _, v := range line {
            if found == v {
                fmt.Printf("(%c)", found)
            } else {
                fmt.Printf("%c", v)
            }
        }
        fmt.Printf("\nfound: %c\n", found)
        s := get_score(found)
        fmt.Printf("score: %d\n\n", s)
        total += s
    }
    fmt.Printf("total: %d", total)
}

func search(s string) rune {
    l := len(s)

    // split
    a := s[:l/2]
    b := s[l/2:]

    // search
    for _, i := range a {
        for _, j := range b {
            if i == j {
                return i
            }
        }
    }
    return 0
}

func get_score(found rune) int {
    if unicode.IsLower(found) {
        return int(found) - 96
    }
    return int(found) - 38
}
